src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -91,117 +93,190 @@ export class KernelExecution implements IDisposable {
             );
         }
     }
-
-    public async cancelCell(cell: NotebookCell) {
-        const execution = this.cellExecutions.get(cell);
-        if (execution) {
-            this.cellExecutions.delete(cell);
-            traceCellMessage(cell, 'Cancel cell from Kernel Execution');
-            await execution.cancel();
-        } else {
-            traceCellMessage(cell, 'Cannot cancel cell execution from Kernel Execution');
+    public async interrupt(notebookPromise: Promise<INotebook>, document: NotebookDocument): Promise<InterruptResult> {
+        const notebook = await notebookPromise;
+        if (!notebook) {
+            traceInfo('No notebook to interrupt');
+            return InterruptResult.Success;
         }
+        traceInfo('Interrupt kernel execution');
+        // Interrupt all the cells, & wait for all to complete.
+        const cellExecutionPromises = this.waitForAllQueuedCells(document, true);
+
+        // Interrupt the active execution
+        const result = this._interruptPromise
+            ? await this._interruptPromise
+            : await (this._interruptPromise = this.interruptExecution(notebook.session, cellExecutionPromises));
+
+        // Done interrupting, clear interrupt promise
+        this._interruptPromise = undefined;
+
+        return result;
+    }
+    public dispose() {
+        this.disposables.forEach((d) => d.dispose());
+    }
+    private async cancelAllCells(notebookPromise: Promise<INotebook>, document: NotebookDocument): Promise<void> {
+        await this.interrupt(notebookPromise, document);
+        await this.waitForAllQueuedCells(document);
     }
 
-    public async cancelAllCells(document: NotebookDocument): Promise<void> {
-        if (this.documentExecutions.get(document)) {
-            this.documentExecutions.get(document)!.cancel();
+    private async waitForAllQueuedCells(document: NotebookDocument, interrupt = false) {
+        traceInfo('Interrupt kernel execution');
+        const stackOfCellsToExecute = this.stackOfCellsToExecuteByDocument.get(document);
+        if (!Array.isArray(stackOfCellsToExecute) || stackOfCellsToExecute.length === 0) {
+            this.stackOfCellsToExecuteByDocument.delete(document);
+            // No pending cells.
+            return;
         }
+
         traceInfo('Cancel document execution');
-        await Promise.all(document.cells.map((cell) => this.cancelCell(cell)));
+        // Check all cells
+        const cellsToCancel = new Set([...document.cells, ...stackOfCellsToExecute.map((item) => item.cell)]);
+        const cellExecutions = Array.from(cellsToCancel)
+            .map((cell) => this.cellExecutions.get(cell))
+            .filter((item) => item !== undefined)
+            .map((item) => item!);
+
+        // Wait for all to complete.
+        await Promise.all(
+            cellExecutions.map(async (item) => {
+                if (interrupt) {
+                    await item.interrupt();
```

Comment:
Now I'm really confused. I thought cellExecution didn't have interrupt anymore?

Reply:
Already on this, I too realized this & used existing `cancel`